test/e2e/kustomizations.go

```diff
@@ -0,0 +1,168 @@
+  path: /spec/template/spec/containers/1/env/-
+  value:
+    name: CONTROLLER_KONG_ADMIN_INIT_RETRIES
+    value: "%d"`
+
+	livenessProbePatch = `- op: replace
+  path: /spec/template/spec/containers/%[1]d/livenessProbe/initialDelaySeconds
+  value: %[2]d
+- op: replace
+  path: /spec/template/spec/containers/%[1]d/livenessProbe/timeoutSeconds
+  value: %[3]d
+- op: replace
+  path: /spec/template/spec/containers/%[1]d/livenessProbe/failureThreshold
+  value: %[4]d`
+)
+
+// getKustomizedManifest takes a base manifest Reader and a Kustomization, and returns the manifest after applying the
+// Kustomization. It overwrites the Kustomization's Bases; any existing Bases will be discarded.
+func getKustomizedManifest(baseManifestReader io.Reader, kustomization types.Kustomization) (io.Reader, error) {
+	workDir, err := os.MkdirTemp("", "kictest.")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(workDir)
+	orig, err := io.ReadAll(baseManifestReader)
+	if err != nil {
+		return nil, err
+	}
+	err = os.WriteFile(filepath.Join(workDir, "base.yaml"), orig, 0o600)
+	if err != nil {
+		return nil, err
+	}
+	kustomization.Bases = []string{"base.yaml"}
+	marshalled, err := yaml.Marshal(kustomization)
+	if err != nil {
+		return nil, err
+	}
+	err = os.WriteFile(filepath.Join(workDir, "kustomization.yaml"), marshalled, 0o600)
+	if err != nil {
+		return nil, err
+	}
+	kustomized, err := runKustomize(workDir)
+	if err != nil {
+		return nil, err
+	}
+	return bytes.NewReader(kustomized), nil
+}
+
+// patchControllerImage replaces the kong/kubernetes-ingress-controller image with the provided image and tag,
+// and returns the modified manifest.
+func patchControllerImage(baseManifestReader io.Reader, image, tag string) (io.Reader, error) {
+	kustomization := types.Kustomization{
+		Bases: []string{"base.yaml"},
+		Images: []types.Image{
+			{
+				Name:    "kong/kubernetes-ingress-controller",
+				NewName: image,
+				NewTag:  tag,
+			},
+		},
+	}
+	return getKustomizedManifest(baseManifestReader, kustomization)
+}
+
+// patchKongImage replaces the kong and kong/kong-gateway images in a manifest with the provide image and tag,
+// and returns the modified manifest.
+func patchKongImage(baseManifestsReader io.Reader, image, tag string) (io.Reader, error) {
+	kustomization := types.Kustomization{
+		Bases: []string{"base.yaml"},
+		Images: []types.Image{
+			{
+				Name:    "kong/kong-gateway",
+				NewName: image,
+				NewTag:  tag,
+			},
+			{
+				Name:    "kong",
+				NewName: image,
+				NewTag:  tag,
+			},
+		},
+	}
+	return getKustomizedManifest(baseManifestsReader, kustomization)
+}
+
+// patchControllerStartTimeout adds or updates the controller container CONTROLLER_KONG_ADMIN_INIT_RETRIES and
+// CONTROLLER_KONG_ADMIN_INIT_RETRY_DELAY environment variables with the provided values, and returns the modified
+// manifest.
+func patchControllerStartTimeout(baseManifestReader io.Reader, tries int, delay string) (io.Reader, error) {
```

Comment:
... and here on the other hand `delay` is expressed as `string`? Why the difference with `period` using `int` below? Can we also use `time.Duration` for `delay` here?

Reply:
My design for both approaches it as if you're editing the actual manifests (that's the process I'm working from), so the types are the types from the original JSON/YAML.

The two are targeting separate pieces of configuration that chose different representation of time durations:
- The Kubernetes livenessProbe type uses seconds only for all its durations, so those got ints.
- The controller `--kong-admin-init-retry-delay` uses a pflag duration, so the strings understood by https://pkg.go.dev/time#ParseDuration

The other option was to make the arg a single dedicated type partially modeling the patched resource, I just didn't bother since I figured we wouldn't look at these often; they're single-purpose wrappers whose values we hopefully don't need to adjust once they make E2E happy.

I kinda prefer the "it's the same value you stick in the manifest" approach but don't really care. Update makes them the native Go types. LMKWYT

Reply:
Someone once said: you write code once but will read it multiple times.

I get what you mean by using the same (as in verbatim) values as in the manifests but OTOH these are not manifests: these are Go functions that can (and should) leverage the type system to ease the reader's mental burden.

So I'd much rather see us use appropriate types instead as suggested before.

Reply:
The reason I'm advocating this change is to make call sites like these:

```
patchLivenessProbes(manifestsReader, 1, 15, 3, 10)
```

more understandable at first glance.